rasa/core/policies/intent_ted_policy.py

```diff
@@ -0,0 +1,690 @@
+            tracker, domain, interpreter
+        )
+
+        model_data = self._create_model_data(tracker_state_features)
+        output = self.model.run_inference(model_data)
+
+        # take the last prediction in the sequence
+        similarities = output["similarities"][:, -1, :]
+
+        # Check for unlikely intent
+        query_intent = tracker.get_last_event_for(UserUttered).intent_name
+        is_unlikely_intent = self._check_unlikely_intent(
+            domain, similarities, query_intent
+        )
+
+        confidences = list(np.zeros(domain.num_actions))
+
+        if is_unlikely_intent:
+            confidences[domain.index_for_action(ACTION_UNLIKELY_INTENT_NAME)] = 1.0
+
+        return self._prediction(
+            confidences,
+            action_metadata=self._collect_action_metadata(domain, similarities),
+        )
+
+    def _should_check_for_intent(self, intent: Text, domain: Domain) -> bool:
+        """Check if the intent should raise `action_unlikely_intent`.
+
+        Args:
+            intent: Intent to be queried.
+            domain: Domain of the assistant.
+
+        Returns:
+            Whether intent should raise `action_unlikely_intent` or not.
+        """
+        if domain.intents.index(intent) not in self.label_thresholds:
+            # This means the intent was never present in a story
+            logger.debug(
+                f"Query intent index {domain.intents.index(intent)} not "
+                f"found in label thresholds - {self.label_thresholds}"
+            )
+            return False
+        if intent in self.config[IGNORE_INTENTS_LIST]:
+            return False
+
+        return True
+
+    def _check_unlikely_intent(
+        self, domain: Domain, similarities: np.array, query_intent: Text
+    ) -> bool:
+        """Check if the query intent is probable according to model's predictions.
+
+        If the similarity prediction for the intent of
+        is lower than the threshold calculated for that
+        intent during training, the corresponding user
+        intent is unlikely.
+
+        Args:
+            domain: Domain of the assistant.
+            similarities: Predicted similarities for all intents.
+            query_intent: Intent to be queried.
+
+        Returns:
+            Whether query intent is likely or not.
+        """
+        predicted_intent_scores = {
+            index: similarities[0][index] for index, intent in enumerate(domain.intents)
+        }
+        sorted_intent_scores = sorted(
+            [
+                (intent_label, score)
+                for intent_label, score in predicted_intent_scores.items()
+            ],
+            key=lambda x: x[1],
+        )
+        query_intent_id = domain.intents.index(query_intent)
+        query_intent_similarity = similarities[0][query_intent_id]
+
+        logger.debug(f"Querying for intent {query_intent}")
+
+        if self._should_check_for_intent(query_intent, domain):
```

Comment:
How come this doesn't happen earlier?

Reply:
Where would you like to place it?

Reply:
I would put the below at the very beginning of the method. Otherwise you're creating  variables like `sorted_intent_scores` even if they may not be needed.
```
if self._should_check_for_intent(query_intent, domain):
    return False
```

Reply:
You mean `if not ...`, right?

Reply:
Yep, oops

Reply:
Good point! Refactored it to have the check at the top.